pallets/hsm/src/lib.rs

```diff
@@ -0,0 +1,1388 @@
+/// and there can be only one asset from a single pool -3 should be enough.
+/// That means that we would have to have 3 pools with Hollar and some different assets.
+pub const MAX_COLLATERALS: u32 = 3;
+
+/// Unsigned transaction priority for arbitrage
+pub const UNSIGNED_TXS_PRIORITY: u64 = 100;
+
+/// Offchain Worker lock
+pub const OFFCHAIN_WORKER_LOCK: &[u8] = b"hydradx/hsm/arbitrage-lock/";
+/// Lock timeout in milliseconds
+pub const LOCK_TIMEOUT: u64 = 5_000; // 5 seconds
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use pallet_evm::GasWeightMapping;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_stableswap::Config + SendTransactionTypes<Call<Self>>
+	where
+		<Self as frame_system::Config>::AccountId: AsRef<[u8; 32]> + IsType<AccountId32>,
+	{
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Asset ID of Hollar
+		#[pallet::constant]
+		type HollarId: Get<Self::AssetId>;
+
+		/// Pallet ID to determine HSM account
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// GHO contract address - EVM address of GHO token contract
+		type GhoContractAddress: BoundErc20<AssetId = Self::AssetId>;
+
+		/// Currency - fungible tokens trait to access token transfers
+		type Currency: Mutate<Self::AccountId, Balance = Balance, AssetId = Self::AssetId>;
+
+		/// EVM handler
+		type Evm: EVM<types::CallResult>;
+
+		/// EVM address converter
+		type EvmAccounts: InspectEvmAccounts<Self::AccountId>;
+
+		/// The gas limit for the execution of EVM calls
+		#[pallet::constant]
+		type GasLimit: Get<u64>;
+
+		/// Gas to Weight conversion.
+		type GasWeightMapping: GasWeightMapping;
+
+		/// Weight information for the extrinsics.
+		type WeightInfo: WeightInfo;
+
+		#[cfg(feature = "runtime-benchmarks")]
+		type BenchmarkHelper: traits::BenchmarkHelper<Self::AccountId>;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	/// List of approved assets that Hollar can be purchased with
+	///
+	/// This storage maps asset IDs to their collateral configuration information.
+	/// Only assets in this map can be used to mint or redeem Hollar through HSM.
+	/// Each collateral has specific parameters controlling its usage in the HSM mechanism.
+	#[pallet::storage]
+	#[pallet::getter(fn collaterals)]
+	pub type Collaterals<T: Config> = StorageMap<_, Blake2_128Concat, T::AssetId, CollateralInfo<T::AssetId>>;
+
+	/// Amount of Hollar bought with an asset in a single block
+	///
+	/// This storage tracks how much Hollar has been bought back by HSM for each collateral
+	/// asset within the current block. This is used to enforce rate limiting on Hollar redemptions.
+	/// Values are reset to zero at the end of each block in on_finalize.
+	#[pallet::storage]
+	#[pallet::getter(fn hollar_amount_received)]
+	pub type HollarAmountReceived<T: Config> = StorageMap<_, Blake2_128Concat, T::AssetId, Balance, ValueQuery>;
```

Comment:
is there actually reason this is per collateral and not total? having total would simplify it and number of collaterals accepted could be unlimited completely

Reply:
it is basically based on different pool states, so from one pool, you can accept more etc. that menas perasset makes sense,